manifold/src/csg_tree.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+#include "manifold.h"
+
+namespace manifold {
+
+enum class CsgNodeType { UNION, INTERSECTION, DIFFERENCE, LEAF };
+
+class CsgLeafNode;
+
+class CsgNode {
+ public:
+  virtual std::shared_ptr<CsgLeafNode> ToLeafNode() const = 0;
+  virtual std::shared_ptr<CsgNode> Transform(const glm::mat4x3 &m) const = 0;
+  virtual CsgNodeType GetNodeType() const = 0;
+  virtual glm::mat4x3 GetTransform() const = 0;
+
+  std::shared_ptr<CsgNode> Translate(const glm::vec3 &t) const;
+  std::shared_ptr<CsgNode> Scale(const glm::vec3 &s) const;
+  std::shared_ptr<CsgNode> Rotate(float xDegrees = 0, float yDegrees = 0,
+                                  float zDegrees = 0) const;
+};
+
+class CsgLeafNode final : public CsgNode {
+ public:
+  CsgLeafNode();
+  CsgLeafNode(std::shared_ptr<const Manifold::Impl> pImpl_);
+  CsgLeafNode(std::shared_ptr<const Manifold::Impl> pImpl_,
+              glm::mat4x3 transform_);
+
+  std::shared_ptr<const Manifold::Impl> GetImpl() const;
+
+  Box GetBoundingBox() const;
+
+  std::shared_ptr<CsgLeafNode> ToLeafNode() const override;
+
+  std::shared_ptr<CsgNode> Transform(const glm::mat4x3 &m) const override;
+
+  CsgNodeType GetNodeType() const override;
+
+  glm::mat4x3 GetTransform() const override;
+
+  static Manifold::Impl Compose(
+      const std::vector<std::shared_ptr<CsgLeafNode>> &nodes);
```

Comment:
I'm fine with this, but I can't do any reordering to this because it seems that decompose depends on this and would not work if the operation is replaced with union for example.

Reply:
I don't think we need to worry about that. Technically I don't guarantee anything for input that's self-overlapping anyway, so I don't really consider a Compose of intersecting objects to be valid. Ideally I'll make a function to remove self-overlaps eventually, but that's non-trivial.